Fix server side copy of files with `?` in - fixes #7058

ncw opened the pull request:
## Description

Before this change the CopyObjectHandler and the CopyObjectPartHandler
both looked for a `versionId` parameter on the `X-Amz-Copy-Source` URL
for the version of the object to be copied.  According to S3
documentation the version should be found in the
`X-Amz-Copy-Source-Version-Id` field.

A consequence of this was that files with `?` in their names could not
be server side copied.

This fixes the problem by reading the version ID from the correct
header instead of the URL.

S3 Docs:
- https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectCOPY.html
- https://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadUploadPartCopy.html



## Motivation and Context

See #7058 

## Regression
<!-- Is this PR fixing a regression? (Yes / No) -->
<!-- If Yes, optionally please include minio version or commit id or PR# that caused this regression, if you have these details. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

harshavardhana commented on cmd/object-handlers.go:
```diff
@@ -675,6 +671,21 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
+
+	// Unescape Copy source path if necessary.
+	cpSrcPathUnescaped, err := url.QueryUnescape(cpSrcPath)
+	if err != nil {
```
Shouldn't this be `if err == nil`

ncw replied:
Yes - well spotted!

harshavardhana commented on cmd/object-handlers.go:
```diff
@@ -1487,6 +1494,21 @@ func (api objectAPIHandlers) CopyObjectPartHandler(w http.ResponseWriter, r *htt
+
+	// Unescape Copy source path if necessary.
+	cpSrcPathUnescaped, err := url.QueryUnescape(cpSrcPath)
+	if err != nil {
```
Shouldn't this be `if err == nil` ?

ncw replied:
Yes - well spotted!

ncw committed:
Fix server side copy of files with `?` in - fixes #7058

Before this change the CopyObjectHandler and the CopyObjectPartHandler
both looked for a `versionId` parameter on the `X-Amz-Copy-Source` URL
for the version of the object to be copied on the URL unescaped version
of the header.  This meant that files that had question marks in were
truncated after the question mark so that files with `?` in their
names could not be server side copied.

After this change the URL unescaping is done during the parsing of the
`versionId` parameter which fixes the problem.

This change also introduces the same logic for the
`X-Amz-Copy-Source-Version-Id` header field which was previously
ignored, namely returning an error if it is present and not `null`
since minio does not currently support versions.

S3 Docs:
- https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectCOPY.html
- https://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadUploadPartCopy.html

harshavardhana reviewed (dismissed):
LGTM
